be/src/runtime/buffer_control_block.cpp

```diff
@@ -79,145 +80,162 @@ void GetResultBatchCtx::on_data(TFetchDataResult* t_result, int64_t packet_seq,
     if (_is_cancelled) {
         return Status::Cancelled("Cancelled BufferControlBlock::add_batch");
     }
-
-    int num_rows = result->result_batch.rows.size();
-
-    while ((!_batch_queue.empty() && (num_rows + _buffer_rows) > _buffer_limit) && !_is_cancelled) {
+    // serialize
+    auto status_or_value = _serialize_result(result);
+    RETURN_IF_ERROR(status_or_value.status());
+    auto ser_res = std::move(status_or_value.value());
+    std::unique_lock<std::mutex> l(_lock);
+    while ((_batch_queue.size_approx() > _buffer_limit || _buffer_bytes > _max_memory_usage) && !_is_cancelled) {
         _data_removal.wait(l);
     }
 
     if (_is_cancelled) {
         return Status::Cancelled("Cancelled BufferControlBlock::add_batch");
     }
 
-    if (_waiting_rpc.empty()) {
-        _buffer_rows += num_rows;
-        _batch_queue.push_back(result);
-        _data_arriaval.notify_one();
-    } else {
-        auto* ctx = _waiting_rpc.front();
-        _waiting_rpc.pop_front();
-        ctx->on_data(result, _packet_num);
+    _process_batch_without_lock(ser_res);
+    // should delete it outside in abnormal cases
+    if (need_free) {
         delete result;
-        _packet_num++;
     }
     return Status::OK();
 }
 
-Status BufferControlBlock::add_batch(std::unique_ptr<TFetchDataResult>& result) {
-    std::unique_lock<std::mutex> l(_lock);
-
-    if (_is_cancelled) {
-        return Status::Cancelled("Cancelled BufferControlBlock::add_batch");
-    }
-    int num_rows = result->result_batch.rows.size();
-    while ((!_batch_queue.empty() && (num_rows + _buffer_rows) > _buffer_limit) && !_is_cancelled) {
-        _data_removal.wait(l);
-    }
-    if (_is_cancelled) {
-        return Status::Cancelled("Cancelled BufferControlBlock::add_batch");
+StatusOr<std::unique_ptr<SerializeRes>> BufferControlBlock::_serialize_result(TFetchDataResult* result) {
+    uint8_t* buf = nullptr;
+    uint32_t len = 0;
+    auto ser_res = std::make_unique<SerializeRes>();
+    ser_res->row_size = result->result_batch.rows.size();
+    {
+        SCOPED_TIMER(_rpc_serialize_timer);
+        ThriftSerializer ser(false, 4096);
+        RETURN_IF_ERROR(ser.serialize(&result->result_batch, &len, &buf));
+        ser_res->attachment.append(buf, len);
     }
+    return std::move(ser_res);
+}
 
-    _process_batch_without_lock(result);
-    return Status::OK();
+Status BufferControlBlock::add_batch(std::unique_ptr<TFetchDataResult>& result) {
+    return add_batch(result.get(), false);
 }
 
-void BufferControlBlock::_process_batch_without_lock(std::unique_ptr<TFetchDataResult>& result) {
+void BufferControlBlock::_process_batch_without_lock(std::unique_ptr<SerializeRes>& ser_res) {
     if (_waiting_rpc.empty()) {
-        _buffer_rows += result->result_batch.rows.size();
-        _batch_queue.push_back(result.release());
+        _buffer_bytes += ser_res->attachment.length();
+        _batch_queue.enqueue(std::move(ser_res));
```

Comment:
Should use `ConcurrentQueue::enqueue(producer_token_t, T)` instead of `ConcurrentQueue::enqueue(T)`, which doesn't preserve the order of the input by default.

Reply:
ok, use prevous list as here is hard to know whether data is from sort or not.